File: programmers/완전탐색/전력망을 둘로 나누기.java
// Tree 형태 : Cycle이 존재하지 않는다

class Solution {
    int abs(int a){
        if(a < 0)
            return a * -1;
        return a;
    }
    
    int proc(int[][]map, int[]check, int now, int n){
        int cnt = 1;
        for(int i = 1;i <= n;i++){
            if(map[now][i] == 1 && check[i] != 1){
                check[i] = 1;
                cnt += proc(map, check, i, n);
            }
        }
        return cnt;
    }
    
    public int solution(int n, int[][] wires) {
        int N = wires.length;
        int answer = 987654321;
        
        int[][]map = new int[n + 10][n + 10];
        for(int i = 0;i < N;i++){
            map[wires[i][0]][wires[i][1]] = 1;
            map[wires[i][1]][wires[i][0]] = 1;
        }
        
        for(int i = 0;i < N;i++){
            int[]check = new int[n + 10];
            map[wires[i][0]][wires[i][1]] = 0;
            map[wires[i][1]][wires[i][0]] = 0;
            
            int now = wires[i][0];
            check[now] = 1;
            
            int cnt = proc(map, check, now, n);
            int leftOver = n - cnt;
            
            answer = Math.min(answer, abs(leftOver - cnt));
            
            map[wires[i][0]][wires[i][1]] = 1;
            map[wires[i][1]][wires[i][0]] = 1;
        }
        
        return answer;
    }
}

File: programmers/완전탐색/전력망을 둘로 나누기.py
from collections import deque
import sys

def topology_count(start, check, map):
    cnt = 0
    q = deque([])
    q.append(start)
    check[start] = 1
    while q:
        now = q.popleft()
        for i in map[now]:
            if check[i] == 1:
                continue
            q.append(i)
            check[i] = 1
            cnt += 1
    return cnt

def solution(n, wires):
    answer = sys.maxsize
    for i in range(len(wires)):
        check = [0 for _ in range(n + 1)]
        map = [[] for _ in range(n + 1)]
        # 무방향 그래프
        for j in range(len(wires)):
            if i == j:
                continue
            map[wires[j][0]].append(wires[j][1])
            map[wires[j][1]].append(wires[j][0])
        topology_1 = topology_count(1, check, map)
        topology_2 = 0
        for i in range(2, n + 1):
            if check[i] == 0:
                topology_2 = topology_count(i, check, map)
                break
        answer = min(answer, abs(topology_1 - topology_2))
    return answer